### ISSUE TITLE
`handlers(global = TRUE)` does not work as expected with Rmd files

### ISSUE THREAD

**Issue by LukasWallrich** [HenrikBengtsson/progressr#129]
Thanks for this very helpful package! I am usually interactively working within Rmd files, and something strange is going on there.

When I run the code below by sourcing an R script, I afterwards consistently get progress updates in the Console. However, when I run the same as a chunk in an Rmd file,  I only get updates from the last line included in the chunk - and no progress when I run further commands in the console. However, running `handlers(global=NA)` in the console returns TRUE in both situations - so I don't understand what's going on here ... would be great to get consistent behaviour?

```
library(progressr)
handlers(global = TRUE)
slow_sum <- function(x) {
  p <- progressr::progressor(along = x)
  sum <- 0
  for (kk in seq_along(x)) {
    Sys.sleep(0.1)
    sum <- sum + x[kk]
    p(message = sprintf("Added %g", x[kk]))
  }
  sum
}
slow_sum(1:10)
```

**Comment by LukasWallrich** [comment]
I experimented a bit further - once I have set `handlers(global = TRUE)`  from the Rmd file, I cannot recover in the console. Running `handlers(global = TRUE)` does not change the situation (no progress there), but `handlers(global = FALSE)` fails with `Error in globalCallingHandlers(NULL) : should not be called with handlers on the stack` ...

**Comment by HenrikBengtsson** [comment]
Hi, thanks for reporting.  First of all, yes, that we get:

```r
html <- knitr::knit(text = "`r progressr::handlers(global = FALSE)`")
#>  |............................................................| 100%
#>   inline R code fragments
#>
#> Error in globalCallingHandlers(NULL) : 
#>   should not be called with handlers on the stack
```

when we try to use `handlers(globals = TRUE)` or `handlers(globals = FALSE)` in a knitr, Rmarkdown, ... document is actually expected.  This is because such documents are processed with active calling handlers, e.g. the code in the code chunks are evaluated with `tryCatch()` and `withCallingHandlers()`.  R itself does not allow us to call `handlers(globals = TRUE/FALSE)` with such, e.g.

```r
> tryCatch({ progressr::handlers(global = FALSE) }, error = identity)
Error in globalCallingHandlers(NULL) : 
  should not be called with handlers on the stack

> tryCatch({ progressr::handlers(global = TRUE) }, error = identity)
Error in globalCallingHandlers(condition = global_progression_handler) : 
  should not be called with handlers on the stack
```

The culprit is that `progressr::handlers(global = TRUE/FALSE)` relies on `globalCallingHandlers()`. Here is the reason for that error with plain R code:

```r
> tryCatch({ globalCallingHandlers(foo = identity) }, error = identity)
Error in globalCallingHandlers(foo = identity) : 
  should not be called with handlers on the stack
```

This is neither a bug in **progressr** nor a bug in R itself.  It's due to a conservative design on how global calling handlers should work in R.  If it allowed, there's a risk we'd end up getting weird and unpredictable behavior or messages, warnings, errors and so on.  So, at least for now, _one must not use `progressr::handlers(global = TRUE/FALSE)` in code chunks_.

**Comment by HenrikBengtsson** [comment]
Next, your example did cause me to play around a bit more, and I also did indeed notice odd, broken behaviors around the global handler afterward.  For example, here is a vanilla R session showing that global progress handlers works:

```r
$ R --quiet --vanilla

library(progressr)
handlers(global = TRUE)
handlers(handler_txtprogressbar(clear = FALSE))

handlers(global = NA)
#> [1] TRUE
y <- slow_sum(1:3, message = FALSE)
#>  |==================================================| 100%
```

However, if we do:

```r
html <- knitr::knit(text = "`r progressr::handlers(global = FALSE)`")
#>  |............................................................| 100%
#>   inline R code fragments
#>
#> Error in globalCallingHandlers(NULL) : 
#>   should not be called with handlers on the stack
```

we get the error explained in https://github.com/HenrikBengtsson/progressr/issues/129#issuecomment-997126074.  That is expected.  We also get, as expected, that:

```r
handlers(global = NA)
#> [1] FALSE
```

reports that the global progress handlers has been disabled.  So far so good, but it's actually _not_ disabled - it still reports on progress:

```r
y <- slow_sum(1:3, message = FALSE)
#>  |==================================================| 100%
```

That is really weird, because if we ask R if there are any global calling handlers registered, it says there are none:

```r
handlers <- globalCallingHandlers()
str(handlers)
#>  list()
```

But since we see a progress bar above, it means that they are still handled globally.  _This might actually be a bug in R!_

We can "fix" the invisible handler by calling either

```r
handlers(global = TRUE)
```

or

```r
globalCallingHandlers(NULL)
```

After this, 

```r
> y <- slow_sum(1:3, message = FALSE)
```

is silent (as expected).  And we can re-enable it again, e.g.

```r
handlers(global = TRUE)
y <- slow_sum(1:3, message = FALSE)
#>  |==================================================| 100%
```

I'll investigate more. My best guess is that there is a bug in R around this, but either way, you cannot (=must not) set _global_ progress handler in Rmarkdown documents and alike.

**Comment by HenrikBengtsson** [comment]
Yes, there's a bug in R itself. I've just filed [PR#18257](https://bugs.r-project.org/show_bug.cgi?id=18257).

**Comment by HenrikBengtsson** [comment]
> Yes, there's a bug in R itself. I've just filed [PR#18257](https://bugs.r-project.org/show_bug.cgi?id=18257).

This has been fixed for R 4.2.0 in R devel rev 81489 (2022-01-14), cf. https://github.com/wch/r-source/commit/2d5789b759b14329925c697fc80f255b85134604

Note that it is still not possible to set global calling handlers _inside_ an Rmarkdown document.  It has to be set prior to processing it.  This is explained in https://github.com/HenrikBengtsson/progressr/issues/129#issuecomment-997126074.

**Comment by HenrikBengtsson** [comment]
I've now added a "known-issue" section 'The global progress handler cannot be set everywhere' to the vignette on the original problem reported, cf. https://progressr.futureverse.org/articles/progressr-intro.html#the-global-progress-handler-cannot-be-set-everywhere.
